Specification language tutorial

Piankero opened the pull request:
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

Piankero commented:
@markus2330 could you please take a look at the tutorial even before the build server passes?

I actually never knew why [these](https://github.com/Piankero/libelektra/blob/1af930707a9f1bd60a501dd44a3a5bbb122fba9a/doc/tutorials/specification-language.md#L61-L63) lines were needed when writing a specification. Can you please clarify that so I can add it to the docu.

Piankero commented:
jenkins build libelektra please

Piankero commented:
jenkins build libelektra please

Piankero commented:
jenkins build libelektra please

markus2330 commented:
> I actually never knew why these lines were needed when writing a specification

The link does not highlight anything for me. Can you review your own code on the lines in question?

Piankero commented:
jenkins build libelektra please

markus2330 commented on doc/tutorials/specification-language.md:
````diff
@@ -0,0 +1,96 @@
+could be rewritten in an `ini` file in the following way:
+
+```ini
+[]
````
This is added by the ni plugin to indicate that the metadata below is parentKey.

Piankero replied:
What do you mean by "this is added by the ni plugin"? I actually have to add it on the top of the specification files.

Piankero replied:
The `infos/plugins` metadata is optional?

markus2330 replied:
Ok: then let us say it is needed by the ni plugin.

>The infos/plugins metadata is optional?


yes

markus2330 commented on doc/tutorials/specification-language.md:
````diff
@@ -0,0 +1,96 @@
+
+```ini
+[]
+mountpoint = test-specification.ni
````
Is described in detail in http://book.libelektra.org/

markus2330 commented on doc/tutorials/specification-language.md:
```diff
@@ -0,0 +1,97 @@
+## Specification writing
+
+There are multiple ways how you can set constraints for configuration settings. All possibilities though set `metadata` to certain
+configuration settings (this is also done in the spec namespace). A metadata is just additional data describing the configuration setting.
```
Avoid subsentences in (). Here the text in () is imho essential as spec is exactly the way as it should be done.

markus2330 commented on doc/tutorials/specification-language.md:
```diff
@@ -0,0 +1,97 @@
+configuration settings (this is also done in the spec namespace). A metadata is just additional data describing the configuration setting.
+One way to set metadata is the command line tool [kdb meta-set](../help/kdb-meta-set.md) which we have used in the upper example.
+For the [validation plugin](../../src/plugins/validation/README.md) which supports regular expression checks there exists even its own
+command line tool [kdb vset](../help/kdb-vset.md).
```
Actually, I think we should remove `kdb vset` as it does not properly put meta-data to spec. Can you do this?

Piankero replied:
Yes

markus2330 commented on doc/tutorials/specification-language.md:
```diff
@@ -0,0 +1,97 @@
+
+If you are writing validation for dozens of configuration settings, the command line might not be the most convenient way. You can also
+write specifications in any file format of your desire. Examples can be seen [here](../../examples/spec). In Elektra, all file formats
+can have metadata but some needed extra tinkering like any `yaml` parser since metadata are not natively supported. We recommend the `ini`
```
According the docu yamlcpp supports meta data.

Here you actually use the `ni` plugin, not the `ini` plugin. And yes, I would also recommend the `ni` plugin, until we get the TOML plugin.

Piankero replied:
I am referring to the `ini` file format, not the `ini` plugin.

markus2330 replied:
There are many different variants of the INI file format, so the text is not clear.

Furthermore, we usually write:

- INI file format
- `ìni` plugin

markus2330 commented on doc/tutorials/specification-language.md:
````diff
@@ -0,0 +1,97 @@
+type = boolean
+```
+
+To integrate a configuration into the key database, we mount it by using the property `mountpoint` in the configuration specification.
````
What is a property?

Piankero replied:
Would "configuration key" be more appropriate for you?

markus2330 replied:
It should be either explained in the tutorial or in doc/help/elektra-glossary.md

markus2330 commented on doc/tutorials/specification-language.md:
````diff
@@ -0,0 +1,97 @@
+```
+
+The `spec-mount` command tells Elektra to load all relevant plugins. It is smart enough to automatically detect the needed plugins
+by looking at all metadata. You might need to remove some metadata to not load certain plugins that cause the exceeding of the maximum
````
Describe here which plugins it will detect and use in this case.

markus2330 commented on doc/tutorials/specification-language.md:
```diff
@@ -0,0 +1,97 @@
+
+What remains to know are the metadata names and meanings. We have used the `type` plugin which provides the metadata `type` metadata along with
+many other metadata such as `check/enum`. You can read all available metadata in the plugin READMEs which all have to have a
+so called `contract` at the beginning of each file. This contract also contains the `infos/metadata` data which lists all available
```
Continue here with the example of the type plugin. Also give another example, I do not think this text makes the concept "contract" clear.

markus2330 commented on doc/tutorials/specification-language.md:
```diff
@@ -0,0 +1,97 @@
+Many plugins write metadata with arrays in them such as the `type` plugin which allows to list all enumerations for configuration settings.
+We have a nice [tutorial](../tutorials/arrays.md) for you to get used to arrays.
+
+You can also set validation for multiple configurations settings at once by using [globbing](../../src/plugins/glob/README.md).
```
This plugin works only on contract level (between plugins) but not in specifications.

In specifications we have different kind of globbing with `_` and `#` implemented by the spec plugin.

markus2330 commented on doc/tutorials/specification-language.md:
```diff
@@ -0,0 +1,97 @@
+
+You can also set validation for multiple configurations settings at once by using [globbing](../../src/plugins/glob/README.md).
+
+If you decide to write metadata via a file (and especially the `ini` format) you should be aware of multiline strings that can cause
```
`ni` should not have this problem.

markus2330 commented on doc/tutorials/specification-language.md:
```diff
@@ -0,0 +1,97 @@
+parsing issues. Descriptions of metadata often take up multiple lines and
+there are multiple `ini` format standards which tell you differently how to write them.
+Elektra also has two ini plugins (`ni` and `ini`) which also handle multilines differently. We recommend the `ni` plugin which also
+works correctly with our provided [examples](../../examples/spec).
```
You did not mention doc/METADATA.md? Please explain how to read and use this file. This is essential for understanding the specification language.

Piankero replied:
> Please explain how to read and use this file.

I guess you mean metadata. *ini*? I added a paragraph concerning that, I hope it correctly reflects what you have in mind

markus2330 replied:
Yes, METADATA.ini. I am not sure if a paragraph is enough to understand what this file is about and how to practically use it. Understanding METADATA.ini is essential for using the specification language.

Piankero replied:
> Understanding METADATA.ini is essential for using the specification language.

I actually never have used this file and are confused why it is so important. I think you will need to explain it to me first so I can add it to the tutorial

markus2330 replied:
It described formally and in detail in book.libelektra.org

In essence METADATA.ini describes the semantic of the specification language.

Piankero replied:
Where exactly you write about this file in the book? I have read through `Abstraction over Embedded and Separated Metadata` but do not think that this helps in this case

markus2330 replied:
page 92ff

Piankero replied:
> page 92ff

I have read through several pages and understand that it is a meta-meta-specification. I though am not sure how to practically use this file. The actually most useful information for me is in the METADATA.ini file itself which tells that you can describe which configuration files in which formats are present on the system. I also assume that one can change what is regarded as a comment symbol but I am not sure what exactly it can achieve otherwise.

Chemin1 replied:
Would a small hint to the book be interesting here? If yes, then we really need to specify which sites are relevant. I would not have found it without the hint to 92ff.

markus2330 replied:
Currently, it is meant to be read from beginning to end. I plan to rewrite the book to a scriptum so that is more structured by topics but it will need at least one more year to finish this.

Piankero commented on doc/tutorials/specification-language.md:
````diff
@@ -0,0 +1,97 @@
+
+```shell script
+kdb meta-set spec/tests/bool 'type' boolean
+kdb set /tests/bool Batman
````
Does the shell recorder also work for non-connected code snippets?

markus2330 replied:
@sanssecours can you help out here? (Maybe update the docu of shell recorder)

sanssecours replied:
> Does the shell recorder also work for non-connected code snippets?

In theory it might work. I never tried it myself though.

> @sanssecours can you help out here? (Maybe update the docu of shell recorder)

Sorry, I do not think that I have time to do that in the moment.

Piankero replied:
If it works, there would an extra "cleanup" section be needed where you unmount everything and cleanup all created tickets. This though is more confusing for a new reader and just pumps up the tutorial in terms of lines. I would prefer to spare that and execute everything local before merging it into master to avoid typos

markus2330 replied:
This cleanup also needs to be done on the computer where a reader typed in the commands, so I do not see any problems with having this.

Local verification if a tutorial works, however, has many problems and it was often shown that it does not work because bugs in tutorials usually only lead to frustrations and not to bug reports, e.g. see #3207

Chemin1 replied:
@Piankero You can take a look at doc/tutorials/install-config-files.md

> Does the shell recorder also work for non-connected code snippets?

Yes, I did this in install-config-files.

>  there would an extra "cleanup" section be needed where you unmount everything and cleanup all created keys

Correct and it is a bit annoying to get this one correct. You have to be very careful that you don't start your test runs in an environment that is messed up from previous tests.

When I spent so much time on the cleanup only I felt stupid, but just yesterday my own shell recorder tests saved me from committing a massive behavior change for the merge without noticing.

Piankero commented on doc/tutorials/specification-language.md:
````diff
@@ -0,0 +1,97 @@
+# Validation Semantic: The value 'Batman' of key 'user/tests/bool' could not be converted into a boolean
+```
+
+Please note as of Elektra v0.9 you cannot add an unlimited amount of plugins. You can see if this issue
````
Honestly I find this information very important because when I started first I even opened an own issue for that despite there was already one present. You run into this error quite fast

markus2330 replied:
I agree. This tutorial should link to the ni plugin and the ni plugin documentation should write about how to make meta-data for parent keys. @ulrikeschaefer also had this problem yesterday.

Piankero commented:
I incorporated all review comments, feel free to check again @markus2330

markus2330 commented on doc/tutorials/specification-language.md:
````diff
@@ -0,0 +1,126 @@
+# Validation Semantic: Value '12' of key 'user/tests/abc' not within range 1-10
+```
+
+You can also find all available metadata in the [METADATA.ini](/doc/METADATA.ini) file. It shows which
````
"also"? Where else? Also mount METADATA.ini and make a shell recorder example with ipaddr.

Piankero replied:
> "also"? Where else?

In the README.md files of every plugin

> Also mount METADATA.ini and make a shell recorder example with ipaddr.

I do not understand exactly what this is supposed to achieve.

markus2330 replied:
> I do not understand exactly what this is supposed to achieve.

To understand that meta-meta-data can also be queried within Elektra. Furthermore, the shell recorder tests will wake us up if one day the semantics of check/ipaddr are changed.

markus2330 commented:
> I incorporated all review comments, feel free to check again @markus2330

There is still no shell recorder integration, it took a few seconds to find bugs in the commands.

Chemin1 commented:
You might want to rebase to or merge from master. The Cirrus errors here have already been adressed in #3193 and #3198

Piankero commented:
@markus2330 I have added the shellrecorder test and incorporated all your desired review changes except for the practical metadata.ini explanation. I would need a tutorial myself to understand that. Could you add this section my branch possibly?

markus2330 commented:
The build server still rejects something. You can ask me questions but it is not the idea that I write it for you.

Piankero commented:
> You can ask me questions but it is not the idea that I write it for you.

Alright, then I'll probably have to ask a lot of questions ;)

You wrote that if the semantics change of a metadata, the shell recorder tests will wake us up because of the metadata.ini. Can you give a technical example of how this is achieved by e.g., assuming that the `ipaddr` now suddenly is something different?
How exactly can you query metadata so that a user actually has a benefit? Is it good if the user "searches" for some sort of functionality (e.g., by a lookup for "ipaddr"?) or is the main purpose something different? Can you give me a full example in a shell recorder test so I can actually understand it in practice? What could a user ask or search for that is only present in the metadata.ini? Can the user severely change the behavior of elektra if he changes the metadata.ini himself? If yes, can you give me an example please. I assume one needs to mount the metadata.ini and perform some operations on the loaded keyset.

I hope that I understand the metadata.ini fully afterwards

markus2330 commented:
> You wrote that if the semantics change of a metadata, the shell recorder tests will wake us up because of the metadata.ini. Can you give a technical example of how this is achieved by e.g., assuming that the ipaddr now suddenly is something different?

If you mount doc/METADATA.ini and make `kdb meta-ls .../check/ipaddr type` then we will get as output ipv4 and ipv6. If METADATA.ini changes, we get a different output.

Btw. doc/METADATA.ini has wrong text at the beginning: we should now recommend to use the ni plugin (as ini gets obsolete) and the script''s path is now scripts/kdb/mount-info instead of scripts/mount-info

> How exactly can you query metadata so that a user actually has a benefit? Is it good if the user "searches" for some sort of functionality (e.g., by a lookup for "ipaddr"?) or is the main purpose something different? 

Not only users can query for this but also applications. E.g. a GUI can make a selector that ipaddr can be either ipv4 or ipv6.

> Can you give me a full example in a shell recorder test so I can actually understand it in practice? 

This again sounds like doing homework for you :wink:

> Can the user severely change the behavior of elektra if he changes the metadata.ini himself?

As no plugin currently reads from there, it is currently not possible to change the behavior simply by modifying METADATA.ini.
